static/app/components/smartSearchBar/index.tsx

```diff
@@ -438,6 +440,103 @@ class SmartSearchBar extends Component<Props, State> {
     }
   }
 
+  runQuickAction = (action: QuickAction) => {
+    const {query} = this.state;
+    const token = this.cursorToken ?? undefined;
+
+    const filterTokens = this.filterTokens;
+
+    if (!action.canRunAction || action.canRunAction(token, this.filterTokens.length)) {
+      switch (action.actionType) {
+        case QuickActionType.Delete: {
+          if (token && filterTokens.length > 0) {
+            const index = filterTokens.findIndex(tok => tok === token) ?? -1;
+
+            const newQuery =
+              // We trim to remove any remaining spaces
+              query.slice(0, token.location.start.offset).trim() +
+              (index > 0 && index < filterTokens.length - 1 ? ' ' : '') +
+              query.slice(token.location.end.offset).trim();
+            this.updateQuery(newQuery);
+          }
+
+          break;
+        }
+        case QuickActionType.Negate: {
+          if (token && token.type === Token.Filter) {
+            if (token.negated) {
+              const newQuery =
+                query.slice(0, token.location.start.offset) +
+                query.slice(token.key.location.start.offset);
+              this.updateQuery(newQuery, this.cursorPosition - 1);
+            } else {
+              const newQuery =
+                query.slice(0, token.key.location.start.offset) +
+                '!' +
+                query.slice(token.key.location.start.offset);
+              this.updateQuery(newQuery, this.cursorPosition + 1);
+            }
+          }
+          break;
+        }
+        case QuickActionType.Next: {
+          let offset = this.state.query.length;
+
+          if (this.searchInput.current && filterTokens.length > 0) {
+            this.searchInput.current.focus();
+
+            if (token) {
+              const tokenIndex = filterTokens.findIndex(tok => tok === token);
+              if (typeof tokenIndex !== 'undefined') {
+                if (tokenIndex + 1 < filterTokens.length) {
+                  offset = filterTokens[tokenIndex + 1].location.end.offset;
+                }
+              }
+            }
+
+            if (this.cursorPosition === this.state.query.length) {
+              offset = filterTokens[0].location.end.offset;
+            }
+
+            this.searchInput.current.selectionStart = offset;
+            this.searchInput.current.selectionEnd = offset;
+            this.updateAutoCompleteItems();
+          }
+
+          break;
+        }
+        case QuickActionType.Previous: {
+          let offset = this.state.query.length;
```

Comment:
seems like this quick action is very similar to the "next" quick action. Could we generalize these both into a function that gets the next token offset in a given token list, e.g. in this case it would be the reversedTokenList